river/Output.zig

```diff
@@ -149,157 +153,50 @@ pub fn sendViewTags(self: Self) void {
-
-    var arena = std.heap.ArenaAllocator.init(util.gpa);
-    defer arena.deinit();
-
-    // Assemble command
-    const layout_command = try std.fmt.allocPrint0(&arena.allocator, "{} {} {} {d} {} {}", .{
-        self.layout,
-        visible_count,
-        self.main_count,
-        self.main_factor,
-        layout_width,
-        layout_height,
-    });
-    const cmd = [_:null]?[*:0]const u8{ "/bin/sh", "-c", layout_command, null };
-    const stdout_pipe = try std.os.pipe();
-
-    const pid = try std.os.fork();
-    if (pid == 0) {
-        std.os.dup2(stdout_pipe[1], std.os.STDOUT_FILENO) catch c._exit(1);
-        std.os.close(stdout_pipe[0]);
-        std.os.close(stdout_pipe[1]);
-        std.os.execveZ("/bin/sh", &cmd, std.c.environ) catch c._exit(1);
-    }
-    std.os.close(stdout_pipe[1]);
-    const stdout = std.fs.File{ .handle = stdout_pipe[0] };
-    defer stdout.close();
-
-    // TODO abort after a timeout
-    const ret = std.os.waitpid(pid, 0);
-    if (!std.os.WIFEXITED(ret.status) or std.os.WEXITSTATUS(ret.status) != 0)
-        return LayoutError.BadExitCode;
-
-    const buffer = try stdout.inStream().readAllAlloc(&arena.allocator, 1024);
-
-    // Parse layout command output
-    var view_boxen = std.ArrayList(Box).init(&arena.allocator);
-    var parse_it = std.mem.split(buffer, "\n");
-    while (parse_it.next()) |token| {
-        if (std.mem.eql(u8, token, "")) break;
-        var box = try parseBox(token);
-        box.x += self.usable_box.x + xy_offset;
-        box.y += self.usable_box.y + xy_offset;
-
-        if (box.width > delta_size) box.width -= delta_size;
-        if (box.height > delta_size) box.height -= delta_size;
-
-        try view_boxen.append(box);
-    }
-
-    if (view_boxen.items.len != visible_count) return LayoutError.WrongViewCount;
-
-    // Apply window configuration to views
-    var i: u32 = 0;
-    var view_it = ViewStack(View).iter(self.views.first, .forward, self.pending.tags, arrangeFilter);
-    while (view_it.next()) |view| : (i += 1) {
-        view.pending.box = view_boxen.items[i];
-        view.applyConstraints();
-    }
-}
-
 fn arrangeFilter(view: *View, filter_tags: u32) bool {
     return !view.destroying and !view.pending.float and
         !view.pending.fullscreen and view.pending.tags & filter_tags != 0;
 }
 
-/// Arrange all views on the output for the current layout. Modifies only
-/// pending state, the changes are not appplied until a transaction is started
-/// and completed.
+/// Try to find a layout for this output and send it a layout demand.
+///
+/// If no suitable layout has been found, all views will simply retain their
+/// current dimensions. So without any active layouts, river will function
+/// like a simple floating WM.
+///
+///  If a layout has been found, the changes of view dimensions are async
+///  and can not be expected to apply with the next transaction. Therefore
+///  the layout handler will start transaction once the layout is done.
```

Comment:
If there is a layout client with the proper namespace, we should not start the transaction until after the client has answered our layout demand.

Reply:
That is, the order of events should look something like this:

1. startTransaction() is called
2. layout_demand event is sent to the layout client
3. layout client responds with dimensions for all views
4. configure events are sent to all views needing them
5. after all views have acked the configure and committed, commitTransaction() is called.

Reply:
I am pretty sure that is what the code does, the comment is just not up-to-date yet.